includes/api/ApiQueryAllUsers.php

```diff
@@ -0,0 +1,383 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ * http://www.gnu.org/copyleft/gpl.html
+ *
+ * @file
+ */
+use Wikia\Util\PerformanceProfilers\UsernameUseProfiler;
+
+/**
+ * Query module to enumerate all registered users.
+ *
+ * @ingroup API
+ */
+class ApiQueryAllUsers extends ApiQueryBase {
+	public function __construct( $query, $moduleName ) {
+		parent::__construct( $query, $moduleName, 'au' );
+	}
+
+	public function execute() {
+		$usernameUseProfiler = new UsernameUseProfiler( __CLASS__, __METHOD__ );
+		$db = $this->getDB();
+		$params = $this->extractRequestParams();
+
+		$prop = $params['prop'];
+		if ( !is_null( $prop ) ) {
+			$prop = array_flip( $prop );
+			$fld_blockinfo = isset( $prop['blockinfo'] );
+			$fld_editcount = isset( $prop['editcount'] );
+			$fld_groups = isset( $prop['groups'] );
+			$fld_rights = isset( $prop['rights'] );
+			$fld_registration = isset( $prop['registration'] );
+			$fld_implicitgroups = isset( $prop['implicitgroups'] );
+		} else {
+			$fld_blockinfo = $fld_editcount = $fld_groups = $fld_registration = $fld_rights = $fld_implicitgroups = false;
+		}
+
+		$limit = $params['limit'];
+
+		$this->addTables( 'user' );
+		$useIndex = true;
+
+		$dir = ( $params['dir'] == 'descending' ? 'older' : 'newer' );
+		$from = is_null( $params['from'] ) ? null : $this->keyToTitle( $params['from'] );
+		$to = is_null( $params['to'] ) ? null : $this->keyToTitle( $params['to'] );
+
+		# MySQL doesn't seem to use 'equality propagation' here, so like the
+		# ActiveUsers special page, we have to use rc_user_text for some cases.
+		$userFieldToSort = $params['activeusers'] ? 'rc_user_text' : 'user_name';
+
+		$this->addWhereRange( $userFieldToSort, $dir, $from, $to );
+
+		if ( !is_null( $params['prefix'] ) ) {
+			$this->addWhere( $userFieldToSort .
+				$db->buildLike( $this->keyToTitle( $params['prefix'] ), $db->anyString() ) );
+		}
+
+		if ( !is_null( $params['rights'] ) ) {
+			$groups = array();
+			foreach( $params['rights'] as $r ) {
+				$groups = array_merge( $groups, User::getGroupsWithPermission( $r ) );
+			}
+
+			$groups = array_unique( $groups );
+
+			if ( is_null( $params['group'] ) ) {
+				$params['group'] = $groups;
+			} else {
+				$params['group'] = array_unique( array_merge( $params['group'], $groups ) );
+			}
+		}
+
+		if ( !is_null( $params['group'] ) && !is_null( $params['excludegroup'] ) ) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Define a constant instead of duplicating this literal "excludegroup" 8 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1192)